storage/database/sharded_database.go

```diff
@@ -155,75 +161,251 @@ func (db *shardedDB) Close() {
+	entries := &entryHeap{}
+	heap.Init(entries)
+	for i, ch := range it.resultChs {
+		if e, ok := <-ch; ok {
+			heap.Push(entries, entryWithShardNum{e, i})
+		}
+	}
+
+chanIter:
+	for len(*entries) != 0 {
+		// check if done
+		select {
+		case <-it.ctx.Done():
+			logger.Trace("[shardedDBIterator] combine worker ends due to ctx")
+			break chanIter
+		default:
+		}
+
+		// look for smallest key
+		minEntry := heap.Pop(entries).(entryWithShardNum)
+
+		// fill resultCh with smallest key
+		it.resultCh <- minEntry.Entry
+
+		// fill used entry with new entry
+		// skip this if channel is closed
+		if e, ok := <-it.resultChs[minEntry.shardNum]; ok {
+			heap.Push(entries, entryWithShardNum{e, minEntry.shardNum})
+		}
+	}
+	logger.Trace("[shardedDBIterator] combine worker finished")
+	close(it.resultCh)
+}
+
+type entryWithShardNum struct {
+	common.Entry
+	shardNum int
+}
+
+type entryHeap []entryWithShardNum
+
+func (e entryHeap) Len() int {
+	return len(e)
+}
+
+func (e entryHeap) Less(i, j int) bool {
+	return bytes.Compare(e[i].Key, e[j].Key) < 0
+}
+func (e entryHeap) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+func (e *entryHeap) Push(x interface{}) {
+	*e = append(*e, x.(entryWithShardNum))
+}
+
+func (e *entryHeap) Pop() interface{} {
+	old := *e
+	n := len(old)
+	element := old[n-1]
+	*e = old[0 : n-1]
+	return element
 }
 
-func (pdi *shardedDBIterator) Next() bool {
-	// TODO-Klaytn implement this later.
-	//var minIter Iterator
-	//minIdx := -1
-	//minKey := []byte{0}
-	//minKeyValue := []byte{0}
-	//
-	//for idx, iter := range pdi.iterators {
-	//	if iter != nil {
-	//		if bytes.Compare(minKey, iter.Key()) >= 0 {
-	//			minIdx = idx
-	//			minIter = iter
-	//			minKey = iter.Key()
-	//			minKeyValue = iter.Value()
-	//		}
-	//	}
-	//}
-	//
-	//if minIter == nil {
-	//	return false
-	//}
-	//
-	//pdi.key = minKey
-	//pdi.value = minKeyValue
-	//
-	//if !minIter.Next() {
-	//	pdi.iterators[minIdx] = nil
-	//}
-	//
+// Next gets the next item from iterators.
+func (it *shardedDBIterator) Next() bool {
+	e, ok := <-it.resultCh
+	if !ok {
+		logger.Debug("[shardedDBIterator] Next is called on closed channel")
+		return false
+	}
+	it.key, it.value = e.Key, e.Val
 	return true
 }
 
-func (pdi *shardedDBIterator) Error() error {
-	// TODO-Klaytn implement this later.
+func (it *shardedDBIterator) Error() error {
+	for i, iter := range it.iterators {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
+				"err", iter.Error(), "shardNum", i, "key", it.key, "val", it.value)
+			return iter.Error()
+		}
+	}
 	return nil
 }
 
-func (pdi *shardedDBIterator) Key() []byte {
-	// TODO-Klaytn implement this later.
-	return nil
+func (it *shardedDBIterator) Key() []byte {
+	return it.key
 }
 
-func (pdi *shardedDBIterator) Value() []byte {
-	// TODO-Klaytn implement this later.
-	return nil
+func (it *shardedDBIterator) Value() []byte {
+	return it.value
+}
```

Comment:
`shardedDBIterator` related functions should be moved upward.

Reply:
review reflected